crates/core/src/db/datastore/system_tables.rs

```diff
@@ -235,231 +248,264 @@ st_fields_enum!(enum StVarFields {
+/// Does not check field types since those aren't included in `StFields` types.
+/// If anything in here is not true, the system is completely broken, so it's fine to assert.
+fn validate_system_table<T: StFields + 'static>(def: &ModuleDef, table_name: &str) {
+    let table = def.table(table_name).expect("missing system table definition");
+    let fields = T::fields();
+    assert_eq!(table.columns.len(), fields.len());
+    for field in T::fields() {
+        let col = table
+            .columns
+            .get(field.col_id().idx())
+            .expect("missing system table field");
+        assert_eq!(&col.name[..], field.name());
+    }
+}
+
+/// See the comment on [`SYSTEM_MODULE_DEF`].
+fn system_module_def() -> ModuleDef {
+    let mut builder = RawModuleDefV9Builder::new();
+
+    // System Table [ST_TABLE_NAME]
+    //
+    // | table_id | table_name  | table_type | table_access |
+    // |----------|-------------|----------- |------------- |
+    // | 4        | "customers" | "user"     | "public"     |
+    let st_table_type = builder.add_type::<StTableRow>();
+    builder
+        .build_table(ST_TABLE_NAME, *st_table_type.as_ref().expect("should be ref"))
+        .with_type(TableType::System)
+        .with_primary_key(StTableFields::TableId)
+        .with_column_sequence(StTableFields::TableId, None)
+        .with_unique_constraint(StTableFields::TableId, None) // Note that unique constraints currently imply indexes.
+        .with_unique_constraint(StTableFields::TableName, None);
+
+    // System Table [ST_COLUMN_NAME]
+    //
+    // | table_id | col_id | col_name | col_type            |
+    // |----------|---------|----------|--------------------|
+    // | 1        | 0       | "id"     | AlgebraicType::U32 |
+    let st_raw_column_type = builder.add_type::<StColumnRow>();
+    builder
+        .build_table(ST_COLUMN_NAME, *st_raw_column_type.as_ref().expect("should be ref"))
+        .with_type(TableType::System)
+        .with_unique_constraint(
+            col_list![StColumnFields::TableId.col_id(), StColumnFields::ColPos.col_id()],
+            None,
+        );
+
+    // System Table [ST_INDEX_NAME]
+    //
+    // | index_id | table_id | index_name  | index_algorithm            |
+    // |----------|----------|-------------|----------------------------|
+    // | 1        |          | "ix_sample" | btree({"columns": [1, 2]}) |
+    let st_index_type = builder.add_type::<StIndexRow>();
+    builder
+        .build_table(ST_INDEX_NAME, *st_index_type.as_ref().expect("should be ref"))
+        .with_type(TableType::System)
+        .with_primary_key(StIndexFields::IndexId)
+        .with_index(
+            v9::RawIndexAlgorithm::BTree {
+                columns: col_list![StIndexFields::IndexId],
+            },
+            "accessor_name_doesnt_matter",
+            None,
+        )
+        .with_column_sequence(StIndexFields::IndexId, None)
+        .with_unique_constraint(StIndexFields::IndexId, None);
+
```

Comment:
They used to be comments like `// TODO: Unique constraint on index name?` that this pr made as it was resolved as `no`. Is this the case? If not, then we need to decide about what to do with names.

Reply:
Ah, I'll just put them back in. We should decide that before 1.0 though.